src/circuit/gadget/sinsemilla/chip.rs

```diff
@@ -0,0 +1,372 @@
+    // of the Sinsemilla hash.
+    generator_table: GeneratorTableConfig,
+    // Fixed column shared by the whole circuit. This is used to load the
+    // x-coordinate of the domain Q, which is then constrained to equal the
+    // initial x_a.
+    constants: Column<Fixed>,
+    // Permutation over all advice columns and the `constants` fixed column.
+    perm: Permutation,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct SinsemillaChip {
+    config: SinsemillaConfig,
+}
+
+impl Chip<pallas::Base> for SinsemillaChip {
+    type Config = SinsemillaConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl SinsemillaChip {
+    pub fn construct(config: <Self as Chip<pallas::Base>>::Config) -> Self {
+        Self { config }
+    }
+
+    pub fn load(
+        config: SinsemillaConfig,
+        layouter: &mut impl Layouter<pallas::Base>,
+    ) -> Result<<Self as Chip<pallas::Base>>::Loaded, Error> {
+        // Load the lookup table.
+        config.generator_table.load(layouter)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    #[allow(non_snake_case)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 5],
+        lookup: (Column<Fixed>, Column<Fixed>, Column<Fixed>),
+        constants: Column<Fixed>,
+        perm: Permutation,
+    ) -> <Self as Chip<pallas::Base>>::Config {
+        let config = SinsemillaConfig {
+            q_sinsemilla1: meta.selector(),
+            q_sinsemilla2: meta.fixed_column(),
+            fixed_y_q: meta.fixed_column(),
+            x_a: advices[0],
+            x_p: advices[1],
+            bits: advices[2],
+            lambda_1: advices[3],
+            lambda_2: advices[4],
+            generator_table: GeneratorTableConfig {
+                table_idx: lookup.0,
+                table_x: lookup.1,
+                table_y: lookup.2,
+            },
+            constants,
+            perm,
+        };
+
+        let two = Expression::Constant(pallas::Base::from_u64(2));
+        // Check that the initial x_A, x_P, lambda_1, lambda_2 are consistent with y_Q.
+        meta.create_gate("Initial y_Q", |meta| {
+            let fixed_y_q = meta.query_fixed(config.fixed_y_q, Rotation::cur());
+
+            // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
+            let Y_A = {
+                let lambda_1 = meta.query_advice(config.lambda_1, Rotation::cur());
+                let lambda_2 = meta.query_advice(config.lambda_2, Rotation::cur());
+                let x_a = meta.query_advice(config.x_a, Rotation::cur());
+
+                // x_r = lambda_1^2 - x_a - x_p
+                let x_r = {
+                    let x_p = meta.query_advice(config.x_p, Rotation::cur());
+                    lambda_1.clone().square() - x_a.clone() - x_p
+                };
```

Comment:
Does it work to move the duplicated code outside the `create_gate` calls?

Reply:
Oh, it's not the same `meta`.

Reply:
It can work, see my review.